### ISSUE TITLE
[HOLD for payment 2023-10-25] [HOLD for payment 2023-10-16] 🟢 [$2000] Get user's location

### ISSUE THREAD

**Issue by hayata-suenaga** [Expensify/App#25855]
_This is a part of the Distance Request project. We're adding a feature with which users can request a reimbursement for gas expenses they incur while traveling by car. Please see the following video to check how this beta feature currently look like. This specific task is related getting the user's location in the flow of creating a distance request._

<details>
<summary>Video</summary>

https://github.com/Expensify/App/assets/98560306/93cbda15-72ed-4bc7-9514-47ac838594fe

</details>

We’ll create a UseCurrentLocationButton which will render a View with styles.flexRow. Inside it will be a pin Expensicon and the text “Use current location”. The UseCurrentLocationButton will be rendered directly below the AddressSearch component.

When the button is pressed we will call [getCurrentPosition(success, error, config)](https://github.com/Expensify/react-native-x-geolocation/blob/e736f1259b67ff54b2c7c8ec0bfbf477e17855a4/src/getCurrentPosition/getCurrentPosition.types.ts#L34) which will automatically prompt for the user’s location if needed. In the success callback we will clear any errors, set the text in the WaypointAddress input to “Your location”, access the user’s current location via position.coords.latitude/longitude, save the waypoint in Onyx with address: ‘Your location’, and then navigate back to the “Request money” page.

_The `getCurrentPosition` function is from the new [`react-native-x-geolocation`](https://www.npmjs.com/package/react-native-x-geolocation) library_

In the error callback we’ll call an action setLocationError() which will set the errorCode on a location object in Onyx. We’re using the errorCode instead of a message because we want to let the component handle displaying the proper error message. We’ll use a new component called LocationErrorMessage to display the error below the button. It will have a similar appearance to the [DotIndicatorMessage](https://github.com/Expensify/App/blob/e7cfb99aa2847b9cdf4467deb3a56ebe258c28cb/src/components/DotIndicatorMessage.js#L35) but it will show one error at a time.

If the error.code is 1 for PERMISSION_DENIED then we will display the message "It looks like you have denied permission to your location. Please [allow location permission in settings](link to settings) and then try again”. We will render this error message using Text components and translation keys for the message before and after the link piece. The text in brackets will use a [TextLink](https://github.com/Expensify/App/blob/e7aae03687588ef2c1df69d39f7a6cf83939762e/src/components/TextLink.js#L37) so that Linking.openURL will be able to open the settings on native and desktop. On web we will link to a new help.expensify.com page where there will be instructions for how to access location settings on each browser. Most modern browsers do not allow pages to link to settings for security reasons.

To fetch the link we’ll add to react-native-x-geolocation a function called getPermissionDeniedHelpLink(helpLink, settingsLink = 'app-settings:LOCATION_SETTINGS'). For the platform specific implementations we simply return the correct link. Although the function is a bit trivial it is configurable from the App side and handles the platform specific piece.

_If I understand correctly, `getPermissionDeniedHelpLink` is a very simple function that returns `helpLink` if the code is running on Web and `settingsLink` if it's running on iOS or Android_

If the error code is 2 for POSITION_UNAVAILABLE or 3 for TIMEOUT then we’ll use DotIndicatorMessage and display an error message like “We were unable to find your location, please try again or enter an address manually”.

<img width="250" alt="Screenshot 2023-08-24 at 7 17 00 AM" src="https://github.com/Expensify/App/assets/98560306/6a4b6c5c-2517-483d-8eb8-b30210161e92">

_[internal link](https://github.com/Expensify/App/issues/22707)_
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0152e9315a07451f08</li>
        <li>Upwork Job ID: 1695128490480984064</li>
        <li>Last Price Increase: 2023-08-25</li>
<li>Automatic offers: </li>
<ul>
<li>huzaifa-99 | Contributor | 26330544</li>
</ul></li>
    </ul>
</details>

**Comment by thienlnam** [comment]
I've gotten started on this but going to let a contributor finish this up so I can focus on some other issues. 

Whoever decides to pick this up can use this if it helps https://github.com/Expensify/App/pull/25868
There are also some pending changes to react-native-x-geolocation that will be merged in this PR https://github.com/Expensify/react-native-x-geolocation/pull/8

**Comment by hayata-suenaga** [comment]
doubling the payout amount

**Comment by b4s36t4** [comment]
Would love to work on this.

**Comment by hayata-suenaga** [comment]
@neonbhai @b4s36t4 @Nikhil-Vats 

I want an initial PR in 24 hours and finish most of the work over the weekend by Monday. Are you available this weeked to work on this?

**Comment by ishpaul777** [comment]
I can work on this @hayata-suenaga

**Comment by b4s36t4** [comment]
Yes, will do.

**Comment by shubham1206agra** [comment]
@hayata-suenaga I would also like to work on this

**Comment by b4s36t4** [comment]
But whose PR will get accepted? I think all the contributors would be will be willing to work, since this is a 2000$ PR.

**Comment by shubham1206agra** [comment]
> But whose PR will get accepted? I think all the contributors would be will be willing to work, since this is a 2000$ PR.

And its not basic like FC migration

**Comment by hayata-suenaga** [comment]
I love how detailed @huzaifa-99 's proposal is. Gonna assign you

**Comment by hayata-suenaga** [comment]
@narefyev91 I assigned a contributor without your approval because this is an urgent issue.

@huzaifa-99 when PR is up, please ping me also 🙇

**Comment by b4s36t4** [comment]
Hey, @hayata-suenaga I think this is kind of odd. @Nikhil-Vats asked about the proposal template is required or not. If you answered maybe you'd be getting more proposals here (maybe good too).

**Comment by hayata-suenaga** [comment]
@parasharrajat could you tell us what you're working on right now?

**Comment by hayata-suenaga** [comment]
@mallenexpensify just for clarification:

An original PR was reverted. The contributor assigned is working on another PR right now

**Comment by mallenexpensify** [comment]
@hayata-suenaga or @narefyev91 can you help summarize where we're at here? Mainly need to know if I'm supposed to pay anyone.  I removed the `Reviewing` label so this'll go overdue so we can do whatever's needed

**Comment by hayata-suenaga** [comment]
@mallenexpensify the first PR was revereted and we're implementing and reviewing the second PR right now

I gonna put the `Reviewig` label back. We don't have to pay anyone at this point until the second PR is merged, deployed, and is waited during the regression period.

**Comment by mallenexpensify** [comment]
ah.... thanks @hayata-suenaga !!!

**Comment by DylanDylann** [comment]
@hayata-suenaga 
![image](https://github.com/Expensify/App/assets/141406735/e9c9b99b-ad41-45c5-9808-708e363be676)

This is the mockup of "use your location" button. This button should display above the suggestion. [PR's](https://github.com/Expensify/App/pull/26546) @huzaifa-99 make this button at the bottom. 
<img width="1440" alt="Screenshot 2023-10-06 at 12 26 53" src="https://github.com/Expensify/App/assets/141406735/cd0efc9a-09fb-418a-b1d7-edefe59bb55b">


Could you help to confirm what is expected

**Comment by trjExpensify** [comment]
The screenshot I pulled there came from the design doc directly, so the design hasn't changed. The example referenced [here](https://github.com/Expensify/App/issues/25855#issue-1865269889) is of a "first time" user where there aren't any recent destinations to display.

**Comment by trjExpensify** [comment]
It's super handy on these PRs for new features to get a review from the design team by adding the `Design` label on the PR.

**Comment by trjExpensify** [comment]
>This is the mockup of "use your location" button. This button should display above the suggestion. https://github.com/Expensify/App/pull/26546 @huzaifa-99 make this button at the bottom.

As for this though when the user start's typing to auto-complete an address, I don't think "use current location" should be shown at all. At that point the user has clearly decided not to use their current location, so it shouldn't be below or above the results. CC: @shawnborton @dannymcclain @dubielzyk-expensify 

So to recap: 

1. When the search field is empty for a first time user, they'll just see "Use current location" 

<img width="251" alt="image" src="https://github.com/Expensify/App/assets/16232057/5414cfe3-ae5a-4abb-b50d-e11deda67160">

P.S ignore the `X` close button in this mockup, it's outdated in new nav. 

2. When the search field is empty for a returning user with recent destinations, they'll see "Use current location" above their recents list 

<img width="253" alt="image" src="https://github.com/Expensify/App/assets/16232057/ea58b81f-1b36-4f21-bc4e-1c8931b5f94c">

P.S ignore the `X` close button in this mockup, it's outdated in new nav. 

3. When the user starts typing in an address, "Use current location" isn't shown at all

<img width="376" alt="image" src="https://github.com/Expensify/App/assets/16232057/4ecbcca9-f352-4c01-a82f-4d9e59f69f48">

**Comment by dubielzyk-expensify** [comment]
I agree that "Use current location" shouldn't show when the user has started typing something. It should only show when the search input is empty like @trjExpensify suggests. The Google Maps example is a great benchmark.

I also agree that we shouldn't ask for location permissions until the user has tapped "Use current location". Given how sensitive permission asking is, we shouldn't throw it up unless we're sure users will tap "Allow".

**Comment by hayata-suenaga** [comment]
cc: @huzaifa-99

**Comment by trjExpensify** [comment]
>I also agree that we shouldn't ask for location permissions until the user has tapped "Use current location". Given how sensitive permission asking is, we shouldn't throw it up unless we're sure users will tap "Allow".

Yup, agreed. We should ask permission until they tap "Use current location" to trigger the OS prompt.

**Comment by dannymcclain** [comment]
Totally agree with @trjExpensify and @dubielzyk-expensify. I think [the screens that Tom posted](https://github.com/Expensify/App/issues/25855#issuecomment-1750394102) are the way to go.

To recap again:
- When the search field is empty for a first time user, they'll see "Use current location" under the input 
- When the search field is empty for a user with recent destinations, they'll see "Use current location" above their recents list
- The OS will only prompt the user for their location _after_ they click "Use current location"
- Once a user starts typing in an address, "Use current location" is not shown.
- So basically:
   - Is the input empty? _Show "Use current location"_
   - Is the input being filled? _Don't show "Use current location"_
   - Did the user tap "Use current location"? _OS location prompt_

**Comment by trjExpensify** [comment]
Ah yes, in production the recent destinations is within a dropdown list. I think we're keeping that as is. I can't quite recall why it was done now, but I believe it was easier that way in the implementation. CC: @neil-marcellini @hayata-suenaga

**Comment by trjExpensify** [comment]
I think it would need to be inside it, otherwise it won't be seen atop the recents list when the search field is auto-focused on pushing to the page.

**Comment by hayata-suenaga** [comment]
waiting for payments

**Comment by mallenexpensify** [comment]
@hayata-suenaga can you help breakdown who is owed what here?  I see three contributors assigned and it looks like there are multiple PRs and one was reverted. 
> @mallenexpensify the first PR was revereted and we're implementing and reviewing the second PR right now

**Comment by hayata-suenaga** [comment]
$2,000 for @huzaifa-99 

and corresponding C+ payments for @parasharrajat and @narefyev91 

@huzaifa-99 did any other C+ reviewed your PRs?

**Comment by hayata-suenaga** [comment]
but this issue is for both the original PR and the follow-up PR right?

did we have another issue and you got paid already?

**Comment by hayata-suenaga** [comment]
Then the payment summary @mallenexpensify provided seems correct 👍 

You did really good job with your PRs and I'm impressed with your perseverance to finish the task despite of its difficulty 🙇

**Comment by mallenexpensify** [comment]
Updated the payment comment above. 
> Contributor: @huzaifa-99 paid $ 2000 via Upwork

@hayata-suenaga since this is new, is there a test we want to add for regressions?   I'm leaning towards no but am unsure.

**Comment by hayata-suenaga** [comment]
It might be quite late to ask this question but I've been always wondering this; "is a regression test something that QA always do to each of every RP or deployment?"

**Comment by mallenexpensify** [comment]
Good question @hayata-suenaga .  My understanding is that there are a suite of regression tests that get run for each deployment. Then, there are also tests that they only run once a month, which is often for design-related issues or edge cases.

**Comment by JmillsExpensify** [comment]
$2,000 payment approved for @parasharrajat based on [this comment](https://github.com/Expensify/App/issues/25855#issuecomment-1787984158).
